### PR TITLE
Add additional information for model error markers

### PR DESCRIPTION
**Pull request by mx990**
Add code, source, and validation-specific data for model error markers. Also avoid checked exceptions when querying marker attributes.

### CODE REVIEW

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java [hunk 1]
```diff
@@ -101,20 +102,21 @@ public static String getTargetUriString(final IResource resource, final EObject
...
 
-	public static Object[] getDiagnosticData(final EObject object, final EStructuralFeature feature) {
+	public static Object[] getDiagnosticData(final EObject object, final EStructuralFeature feature,
+			final String... data) {
 		if (object == null) {
-			return new Object[0];
+			return data;
 		}
-		return new Object[] { object, // [0] object itself (per EMF convention)
+		return Stream.concat(Stream.of(object, // [0] object itself (per EMF convention)
```
Comment (oberlehner) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java:185]: can you add a comment with an example of such string ?

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/ErrorMarkerBuilder.java [hunk 2]
```diff
@@ -70,6 +73,12 @@ public Map<String, Object> getAttributes(final IResource resource) {
 		} else if (target != null) {
 			attributes.put(IMarker.LOCATION, FordiacMarkerHelper.getLocation(target, feature));
 		}
+		if (code != 0) {
```
Comment (oberlehner) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/ErrorMarkerBuilder.java]: no magic numbers here

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacErrorMarker.java [hunk 3]
```diff
@@ -49,24 +47,39 @@ public final class FordiacErrorMarker {
 	private static final Set<String> MODEL_MARKER_TYPES = Set.of(IEC61499_MARKER, VALIDATION_MARKER,
 			INITIAL_VALUE_MARKER, TYPE_DECLARATION_MARKER, IMPORT_MARKER, TYPE_LIBRARY_MARKER, LIBRARY_MARKER);
 
+	public static final String CODE = "org.eclipse.fordiac.ide.model.iec61499.code"; //$NON-NLS-1$
```
Comment (oberlehner) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacErrorMarker.java:60]: what is the difference between code and source ?

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java [hunk 4]
```diff
@@ -130,15 +132,22 @@ public static Map<String, Object> getDiagnosticAttributes(final Diagnostic diagn
 			final List<?> data = diagnostic.getData();
 			if (data != null && data.size() >= 5) {
 				if (data.get(4) != null) {
-					return Map.of(IMarker.LOCATION, data.get(1), FordiacErrorMarker.TARGET_URI, data.get(2),
-							FordiacErrorMarker.TARGET_TYPE, data.get(3), FordiacErrorMarker.TARGET_FEATURE,
-							data.get(4));
+					return Map.of(FordiacErrorMarker.CODE, Integer.valueOf(diagnostic.getCode()),
```
Comment (oberlehner) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java]: can you make a java class or record for this to wrap the data ?
This code is really hard to read.

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/ErrorMarkerBuilder.java (hunk 2 above)
Comment (mx990) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/ErrorMarkerBuilder.java]: Done

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacErrorMarker.java (hunk 3 above)
Comment (mx990) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacErrorMarker.java:60]: The source identifies the particular `EValidator`, see `LibraryElementValidator.DIAGNOSTIC_SOURCE`, and the code is a source-specific error code. I have added additional documentation to make this more clear.

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java (hunk 4 above)
Comment (mx990) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java]: I am also somewhat unhappy with the situation. Unfortunately, I am quite restricted by both the Eclipse platform and Xtext on what types I can use here. The original `Diagnostic.getData()` array must only contain exactly one `EObject` (i.e., the target) and the rest must be plain `String`s to support conversion of the `Diagnostic` to an `org.eclipse.xtext.validation.Issue`. The marker attributes, on the other hand, only support `String`, `Integer`, or `Boolean` as values.

Nevertleless, I tried improving the formatting to (hopefully) make it a bit easier to read.

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java (hunk 1 above)
Comment (mx990) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/errormarker/FordiacMarkerHelper.java:185]: I have added additional documentation on the resulting array.

**Pull request merged**